### DISCUSSION TITLE
Shoud we set TCP_DEFER_ACCEPT on tcp/ssl bind sockets?

### DISCUSSION THREAD

**Discussion by MSP-Greg** [puma/puma#3175] [Ideas]
I believe both Unicorn & Pitchfork set `TCP_DEFER_ACCEPT` on bind sockets.  They set it to 1 seconds as a default, and have a DSL type option to disable it.

It is not supported by all platforms.

Any thoughts?  If enabled, a client connection will only be accepted if it has data.  Or, an 'empty' socket connection won't be accepted.

Not sure about SSL...

**EDIT:** Not sure when/how/why, but I found a note about this in one of my files today...

**Comment by nateberkopec** [comment]
[This StackOverflow answer gives a good overview of why this is useful](https://unix.stackexchange.com/questions/94104/real-world-use-of-tcp-defer-accept)

>Regarding the existance of this option, this is not the traditional behaviour of a socket, normally the socket handler's thread is woken up when the connection is accepted (which is still after the three way handshake completes), and for some protocols activity starts here (e.g. an SMTP server sends a 220 greeting line), but for HTTP the first message in the conversation is the web browser sending its GET/POST/etc line, and until this happens the HTTP server has no interest in the connection (other than timing it out), thus waking up the HTTP process when the socket accept completes is a wasteful activity as the process will immediately fall asleep again waiting for the necessary data.

I think the only real-world performance improvement here would be for clients which connect but never send any data. In this case, `TCP_DEFER_ACCEPT` would prevent Puma from ever seeing that connection, which is nice. In the case where the client is just slow and takes a while to send data, I'm not sure it really matters to Puma due to having the `Reactor`. 

It sounds like a highly marginal gain... worth shipping the extra code and maintaining it? Not sure.
